### PR TITLE
feat(community)_: add version to image url to let clients update

### LINKED ISSUE #16688: [Community] Changing logo doesn't get propagated
Perhaps related (or the same issue):
 
Image
 
If I change the community logo, it's persisted but will update only after app restart (I suspect the banner too but this one is hard to check as it's not visible anywhere else in the app) 

 _Originally posted by @caybro in [#16672](https://github.com/status-im/status-desktop/issues/16672#issuecomment-2451535908)_

### PR DESCRIPTION
**Pull request by jrainville**
Fixes https://github.com/status-im/status-desktop/issues/16688

Since we use the local image server to show the community image, the URL never changes when we update the image, since it's served using a query string containing the community ID. eg: `https://Localhost:46739/communityDescriptionImages?communityID=0x03c5ece7da362d31199fb02d632f85fdf853af57d89c3204b4d1e90c6ec13bb23c&name=thumbnail` Because of that, the clients cannot know if the image was updated, so they had to force update the image every time, which was inefficient.

We discovered this issue when I refactored the community client code in Desktop so that we only update the changed properties of a community instead of reseting the whole thing.

The solution I came up with in the PR is to add a `version` to the URL when we detect that the image changed. This let,s the clients detect when the image was updated without having to do any extra logic.

### CODE REVIEW

File: protocol/communities/community.go
Comment (ilmotta) [inline protocol/communities/community.go]: Maybe a leftover comment?

**Review by ilmotta — APPROVED** [review]

File: server/server_media.go
Comment (caybro) [inline server/server_media.go]: `uint64` here and `uint32` everywhere else?

**Review by caybro — APPROVED** [review]
LGTM

File: protocol/protobuf/chat_identity.proto
Comment (igor-sirotin) [inline protocol/protobuf/chat_identity.proto]: Wait, so `version` is not just local, it becomes part of `CommunityDescription`? 🤔 
Sounds a bit weird to be like that, like this problem should be solved solely on client side, not affecting network messages

Comment (jrainville) [reply protocol/protobuf/chat_identity.proto]: Yeah I agree it should be local only, but I didn't find an easy way to do so, since locally, we use the `IdentityImage` protobuf in the image server.

I'm very open for suggestions to make it locally, but apart from creating some sort of new column or table, I don't know how to make it work.

In the end, this solution is pretty unobtrusive anyway IMO.

Comment (igor-sirotin) [reply protocol/protobuf/chat_identity.proto]: @jrainville I think I'm not convinced, modifying protocol for local needs sounds counter-intuitive to me.

We should probably wrap the `IdentityImage` when passing it to `ServerMedia`, and add the `Version` property to the wrapper. Dunno, or make some method of the media server to update the version of specific image.

> _creating some sort of new column or table_

We don't need to persist the version, it's enough to have it only during a single run. After restart, version can go again from `1`, should work, right?

File: protocol/communities/manager.go
Comment (osmaczko) [inline protocol/communities/manager.go]: This doesn't seem quite right. First, the community is edited, followed by the generation of a new description. Subsequently, the new description undergoes reprocessing, which involves the complete mechanism of decryption and populating encryption-specific cache tables. In this case, the cache is used to store _unencrypted_ descriptions for subsequent processing. However, the invariant is broken in this scenario, and I’m not sure how this ties into the image server.

File: protocol/protobuf/chat_identity.proto
Comment (jrainville) [reply protocol/protobuf/chat_identity.proto]: > We don't need to persist the version, it's enough to have it only during a single run. After restart, version can go again from 1, should work, right?

That's true. I'll try to see if I can do as you say

Comment (osmaczko) [reply protocol/protobuf/chat_identity.proto]: I agree with Igor here.

At first glance, the best would be to use/extend [CommunityChanges](https://github.com/status-im/status-go/blob/ad28f15531c76c216f08118d4055a140ee52f80f/protocol/communities/community_changes.go#L20) with image change recognition and then populate it to client.

**Review by osmaczko — CHANGES_REQUESTED** [review]
Sorry, I need to request changes for this one.

File: protocol/communities/manager.go
Comment (jrainville) [reply protocol/communities/manager.go]: It took me a while to understand what was going on. The image server relies on the community cache to serve the images, ie `encrypted_community_description_cache`.

When calling `EditCommunity` the image does get saved in the `communities` DB. However, the cache does **not** get updated straight away.
From what i could gather, it gets updated the next time `getByID` gets called, which is just a race condition at this point.

Calling `preprocessDescription` after editing the clock makes sure the new description gets saved in the cache DB.

Maybe I could call `SaveDecryptedCommunityDescription` directly instead of `preprocessDescription`? It would be clearer in the code, but `preprocessDescription` does a bit more handling with regards to members, so I thought it would be safer to use that instead.

@osmaczko do you have a suggestion on how to make it better? Maybe a better comment?

File: server/server_media.go
Comment (jrainville) [reply server/server_media.go]: `FormatUint` accepts uint64 only :shrug:

Comment (caybro) [reply server/server_media.go]: :roll_eyes: go doesn't have some sort of generics or method overloads? :open_mouth:

File: protocol/protobuf/chat_identity.proto
Comment (jrainville) [reply protocol/protobuf/chat_identity.proto]: > We don't need to persist the version, it's enough to have it only during a single run. After restart, version can go again from 1, should work, right?

Ok, I tried something simple and I realized that we indeed need to save it somewhere. At least a local cache. However, we do not have a local cache for communities. The closest we have is the `encrypted_community_description_cache` which is a table. It is also what the image server uses to serve the images.

Just to make sure the problem is clear, I can't just pass the updated version to the image server, since the Edit happens in one thread/timeframe and the image serving is a completely different one, so the image server knows nothing about image updates.
It serves using the community cache table, which is populated from the decrypted community description. So we need the version or at least an indicator that the images were updated in there.

So either I create a new local cache for image versions per community, or I add a new column `encrypted_community_description_cache` to save the version, or leave it as I did it. Or you guys have a better option?

File: protocol/communities/manager.go
Comment (osmaczko) [reply protocol/communities/manager.go]: I see. I don't understand why the image server relies on `encrypted_community_description_cache`. The purpose of this cache is tightly coupled with the hash ratchet, meaning the cache is invalidated whenever we receive a key that was previously missing for that community. Therefore, in theory, it wouldn't work correctly if all keys for a given community are received.

This was introduced in the following PR: https://github.com/status-im/status-go/pull/5336. @qfrank, do you recall why the `encrypted_community_description_cache` table was used there? The media server should ideally just use the `GetByID` method (must be injected somehow).

@jrainville would you mind fixing that within the scope of this PR please?

Comment (jrainville) [reply protocol/communities/manager.go]: > @jrainville would you mind fixing that within the scope of this PR please?

I don't mind. I'll just wait for @qfrank 's reply to make sure there isn't something we don,t know about. Maybe the cache serves faster?

Comment (qfrank) [reply protocol/communities/manager.go]: > to make sure there isn't something we don,t know about. Maybe the cache serves faster?

It was easier to implement fetching image through db as media server already contains db instance, that's all :)

Comment (osmaczko) [reply protocol/communities/manager.go]: I understand the reasoning. However, `encrypted_community_description_cache` should not be used to retrieve the community description, as it can be invalidated at any time. This means the entry might be removed and become inaccessible. Moreover, `GetByID` should be used instead because it applies events on top of the raw community description. With the current implementation, if an admin changes the image (logo or banner) and the change is not yet approved by the control node, then the image will not be read properly by the media server.

Comment (jrainville) [reply protocol/communities/manager.go]: Fixed here: https://github.com/status-im/status-go/pull/6127

I'll rebase this PR on top of that other one

**Comment by jrainville** [comment]
@igor-sirotin @osmaczko I rebased this branch on top of https://github.com/status-im/status-go/pull/6127 as I refactored the media sever to not use the community cache.
So I don't need `preprocessDescription` anymore :raised_hands: 

As for the version being in the protobuf, I sadly didn't find a good solution. I'm still hoping you guys can think of something :confused:

**Comment by jrainville** [comment]
Patryk gave me a good suggestion:
> or another trick to do in manager.go:
mediaServer.ImageVersion(func(communityID string) { return imageVersions[communiyID]})

> And in UpdateImageVersions you update the imageVersions map instead ChatIdentity 
This way clients would work fine without changes 

Using a local map in the manager

**Comment by jrainville** [comment]
I updated the PR to go with Patryk's recommendation. I no longer need the protobuf changes and the version isn't saved anymore. Please let me know what you think :smile:

File: protocol/communities/manager.go
Comment (osmaczko) [inline protocol/communities/manager.go]:
```suggestion
			return manager.communityImageVersions[communityID]
```

File: protocol/communities/manager.go
Comment (osmaczko) [inline protocol/communities/manager.go:1634]: Would be great to add a comment explaining why this is needed.

File: server/server_media.go
Comment (osmaczko) [inline server/server_media.go]:
```suggestion
		"version":     {fmt.Sprintf("%d", s.getCommunityImageVersion(communityID))},
```

File: protocol/communities/manager.go
Comment (osmaczko) [inline protocol/communities/manager.go]:
```suggestion
	m.communityImageVersions[communityID] = m.communityImageVersions[communityID] + 1
```

**Review by osmaczko — APPROVED** [review]

**Comment by jrainville** [comment]
FYI @igor-sirotin @osmaczko these two tests seem flaky. They failed on my PR by work locally:
- TestRetrieveBigCommunity
- TestService_IncrementalUpdateMixed

I'll restart them on my PR

**Comment by igor-sirotin** [comment]
> * TestRetrieveBigCommunity
> * TestService_IncrementalUpdateMixed

@jrainville yep they're flaky 💯 
There're some more, appeared all at once 💀

**Comment by osmaczko** [comment]
> FYI @igor-sirotin @osmaczko these two tests seem flaky. They failed on my PR by work locally:
> 
> * TestRetrieveBigCommunity
> * TestService_IncrementalUpdateMixed
> 
> I'll restart them on my PR

You don't need to restart unless one of the tests fails three times in a row, which is highly unlikely even for flaky ones.

**Comment by igor-sirotin** [comment]
> FYI @igor-sirotin @osmaczko these two tests seem flaky. They failed on 
> You don't need to restart unless one of the tests fails three times in a row, which is highly unlikely even for flaky ones.

btw This only works if a test fails. But if it panics, then there's no retries 😐

**Comment by jrainville** [comment]
@igor-sirotin @qfrank @ilmotta can I get one more review please. I updated a lot of the old review wasn't up to date

File: protocol/communities/manager.go
Comment (ilmotta) [inline protocol/communities/manager.go]: Missing `no` word? `So the clients have no way of knowing [...]`

**Review by ilmotta — APPROVED** [review]

**Review by qfrank — APPROVED** [review]
LGTM

BTW, the solution seems pretty specific for community image, I'm wondering would it be better to implement it in a common way for media server, because this issue could also exist with other kind of image, e.g. profile image  🤔

**Comment by caybro** [comment]
> LGTM
> 
> BTW, the solution seems pretty specific for community image, I'm wondering would it be better to implement it in a common way for media server, because this issue could also exist with other kind of image, e.g. profile image 🤔

Yeah, we have exactly the same problem with profile image as well

**Comment by jrainville** [comment]
> LGTM
> 
> BTW, the solution seems pretty specific for community image, I'm wondering would it be better to implement it in a common way for media server, because this issue could also exist with other kind of image, e.g. profile image 🤔

Yeah, when I do the same fix for the profile images, I'll do a pass to generalize the functions, if possible.

But currently on Desktop, only the community image **showed** the issue, because we don't have the "hack" of adding a timestamp to the link each time it changes.

I'll work on that later in the next milestone if I have time

**Pull request merged**
